### DISCUSSION TITLE
UWP -> WinUI3: How is navigating to pages done now?

### DISCUSSION THREAD

**Discussion by myokeeh** [microsoft/TemplateStudio#4584] [Q&A] [answered]
Converting a UWP app. What is the path forward from `ViewModelLocator.cs`?

I was navigating to pages like below. How is that done now?

`ViewModelLocator.Current.NavigationService.Navigate(typeof(temsViewModel).FullName);`

**Comment by mikebattista** [comment, marked as answer]
The `NavigationService` still provides navigation. Instead of `ViewModelLocator.Current`, you can get an instance through dependency injection by adding `INavigationService` to a class constructor, or you can call `App.GetService<INavigationService>()`. Add a `Content Grid` page and look at `ContentGridViewModel` for an example.

**Reply by myokeeh** [reply]
Thanks. Did you mean adding `INavigationAware` to the class constructor? That's what I see on the `ContentGridViewModel`.

Is there a consequence to just using `App.GetService<INavigationService>().NavigateTo(typeof(ItemsViewModel).FullName);` everywhere? That's what I was doing with `ViewModelLocator.Current.NavigationService.Navigate(typeof(ItemsViewModel).FullName);` in `App.xaml.cs` (outside of the page ViewModels).

Also, why isn't this more obvious in the generated pages? Navigating seems like a fairly common thing to need immediately after an app is generated with Template Studio. Could this be added as an option that's a bit more than a Blank Page--call it "Blank Page with Navigation"

**Reply by mikebattista** [reply]
> Is there a consequence to just using App.GetService<INavigationService>().NavigateTo(typeof(ItemsViewModel).FullName); everywhere? That's what I was doing with ViewModelLocator.Current.NavigationService.Navigate(typeof(ItemsViewModel).FullName); in App.xaml.cs (outside of the page ViewModels).

Sure. You can do that.

> Also, why isn't this more obvious in the generated pages? Navigating seems like a fairly common thing to need immediately after an app is generated with Template Studio. Could this be added as an option that's a bit more than a Blank Page--call it "Blank Page with Navigation"

Navigation is handled out-of-the-box with the NavigationView and MenuBar templates. The Blank template is meant to be minimal to give you full control. Blank Page with Navigation would just be more overhead to maintain. If you want a shell for navigation to start with, you should start with the other templates.

**Reply by myokeeh** [reply]
That makes sense. Thanks!
